Product Gallery: Display all Thumbnails and make them scrollable

kmanijak opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

This PR changes the way Product Gallery Thumbnails behave and render:
- Remove "View All" button
- Always render ALL of the Thumbnails
- Make the Thumbnails scrollable
- Change the settings:
    - Remove "No of Thumbnails" control
    - Add "Thumbnail Size" control (percentage, relative to Large image):
        - Min size: 10%
        - Max size: 50%
        - Default size: 33%
- Display gradient indicating more Thumbnails are there (gradient adjust the color to base theme color)

Closes https://github.com/woocommerce/woocommerce/issues/56027
Closes https://github.com/woocommerce/woocommerce/issues/56025

Kudos to @cpapazoglou from whom I took couple of LOC (https://github.com/woocommerce/woocommerce/pull/56124).

TODO / known issue:
- For some reason the Thumbnails shrink when 19+ images are added. Only on Chrome, seems the `flex-basis` is incorrectly accounted for. Interaction makes the block to rerender and fixes the issue
- Mobile layout on Safari looks off

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

#### Sizes - vertical

10% - Editor | 10% - frontend
-- | --
<img width="1022" alt="image" src="https://github.com/user-attachments/assets/972644d7-7527-4fcc-b19c-97e19fcbf587" /> | <img width="1044" alt="image" src="https://github.com/user-attachments/assets/d6783818-99b3-4cff-a1ca-64351626922e" />

50% - Editor | 50% - frontend
-- | --
<img width="1142" alt="image" src="https://github.com/user-attachments/assets/d9945076-ed82-4555-9612-3f1785db9109" /> | <img width="1033" alt="image" src="https://github.com/user-attachments/assets/095b6b81-7d71-4284-82aa-f89a2c35ef82" />

#### Number of images

1 - no Thumbnails | Less than available space | More than available space
-- | -- | --
<img width="1188" alt="image" src="https://github.com/user-attachments/assets/6d4bd83e-e774-4fd6-8c6a-5ddab2b1c676" /> |  <img width="1091" alt="image" src="https://github.com/user-attachments/assets/884b985e-4a28-4142-b680-e24a70c3cd32" /> | <img width="1080" alt="image" src="https://github.com/user-attachments/assets/623266ba-80b2-40ca-9242-a25e7065577f" />

#### Orientation
Previous images show bottom or right gradients (thumbnails are in default state), here I'm scrolling to the "middle" so both gradients are showing up

Vertical | Horizontal
-- | --
<img width="1064" alt="image" src="https://github.com/user-attachments/assets/1635c58d-63fb-48d7-9637-aa6d1e2ad3d2" /> | 
<img width="1143" alt="image" src="https://github.com/user-attachments/assets/79b5a9f2-e87c-46c4-9c15-fe017b063e9b" />

#### Various colour schemes

1 | 2 | 3
-- | -- | --
<img width="1304" alt="image" src="https://github.com/user-attachments/assets/5d3ded7c-099e-4f4e-9d95-b0690ebe1ac7" /> | <img width="1082" alt="image" src="https://github.com/user-attachments/assets/88e9cdbc-0e79-4703-8d1c-9fb3f23d7ddd" /> | <img width="1121" alt="image" src="https://github.com/user-attachments/assets/4e113115-5abf-4616-9314-a64ac0f219bc" />

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Go to the Editor > Single Product template.
2. Replace the Product Image Gallery block with the Product Gallery (beta) block.
3. **Expected:** The preview shows "gradient" at the bottom indicating there's more thumbnail
4. Focus on Thumbnail block, play with the size control and make sure it's reflected in Editor
5. Save and go to the frontend
    - test product with single image - no thumbnails should be displayed
    - test product with 2 images - thumbnails are not scrollable (unless they exceed available space but in most cases they should not with 2 images)
    - test product with 10 images - thumbnails are scrollable and gradient shows up at the bottom and top when there's space to scroll and disappears when you reach the edge

6. Go back to Editor
7. Focus on Gallery Area and transform to Stack - Thumbnails should be horizontal now
<img width="288" alt="image" src="https://github.com/user-attachments/assets/d8ef1d6c-1cce-4f04-8059-e5b968b7aafe" />
8. Repeat steps 3 to 5

10. Go back to Editor
11. Change the global styles color scheme
<img width="284" alt="image" src="https://github.com/user-attachments/assets/c28ceb09-2e65-4c11-a7f2-54bf6ce0d808" />
12. Make sure the gradient adjusts to the base color in editor and frontend (see screenshots above)

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

kmanijak committed:
Remove code related to View All Thumbnails button

kmanijak committed:
Merge branch 'trunk' into add/56025

kmanijak committed:
Remove view all click action

kmanijak committed:
Remove limit of thumbnails

kmanijak committed:
Merge branch 'trunk' into add/56025

kmanijak committed:
Add new thumbnails size attribute

kmanijak committed:
Use the new attribute in the editor

kmanijak committed:
Add new control for percentage size

kmanijak committed:
Adjust the styles

kmanijak committed:
Update the frontend

kmanijak committed:
Apply minor improvements

kmanijak committed:
Adjust the styles to add gradient when there's overflow

kmanijak committed:
Calculate if gradient should be displayed

kmanijak committed:
Display gradient in frontewnd

kmanijak committed:
Adjust the editor

kmanijak committed:
Add new context data

kmanijak committed:
Not sure if migration is needed, probably not

kmanijak committed:
Improve styles

kmanijak committed:
Merge branch 'trunk' into add/56025

kmanijak committed:
Remove View All markup after trunk merging

kmanijak committed:
Pass clicks through the gradient overlay

kmanijak committed:
Fix Editor thumbnails layout

kmanijak committed:
Disable scrolling in the editor

kmanijak committed:
Extract utl to check overflow

kmanijak committed:
Use the util in frontend.tsx

kmanijak committed:
Check the overflow in editor

kmanijak committed:
Add editor styles as gradient conflicts with editor outlines

kmanijak committed:
Reuse styles a bit

kmanijak committed:
Fix linter

kmanijak committed:
Fix linter

kmanijak committed:
Remove obsolete test

kmanijak committed:
Remove migration logic

kmanijak committed:
Display more thumbnails in edit to make sure there's overflow even when size is set to small

kmanijak committed:
Adjust the default to 33%

cpapazoglou commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/block-settings/index.tsx:
```diff
@@ -13,26 +18,27 @@ export const ProductGalleryThumbnailsBlockSettings = ( {
+			__next36pxDefaultSize
 			help={ __(
-				'Choose how many thumbnails (3-8) will display. If more images exist, a “View all” button will display.',
+				'Choose the size of each thumbnail. If thumbnails size is bigger than the product image, thumbnails will turn to slider.',
```
```suggestion
				'Choose the size of each thumbnail in respect to the product image. If thumbnails container size gets bigger than the product image, thumbnails will turn to slider.',
```

cpapazoglou commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/block-settings/index.tsx:
```diff
@@ -13,26 +18,27 @@ export const ProductGalleryThumbnailsBlockSettings = ( {
+			onChange={ ( value: string | undefined ) =>
 				setAttributes( {
-					numberOfThumbnails: Math.round( value ),
+					thumbnailSize: value || '20%',
```
Should this be consistent with the [default value](https://github.com/woocommerce/woocommerce/pull/56069/files#diff-c619be550dcc226ec082db06e8f7ec961c053521c64a211ddb4e3a90ab5df833R16)?
```suggestion
					thumbnailSize: value || '33%',
```

cpapazoglou commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-gallery/frontend.tsx:
```diff
@@ -337,22 +330,15 @@ const productGallery = {
-			const lastThumbnail = thumbnails[ thumbnails.length - 1 ];
-			return context.image.id === lastThumbnail.id;
+			const scrollableElement = getElement()?.ref as HTMLElement;
+			const overflowState = checkOverflow( scrollableElement );
```
Maybe return early if `scrollableElement` is not of type `HTMLElement`?

kmanijak committed:
Force mobile layiout

cpapazoglou commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-gallery/utils.tsx:
```diff
@@ -0,0 +1,26 @@
+export const checkOverflow = ( scrollableElement: HTMLElement ) => {
```
I think we can benefit from typing the returned object too.

kmanijak committed:
Fix lint

cpapazoglou commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-gallery/utils.tsx:
```diff
@@ -0,0 +1,26 @@
+export const checkOverflow = ( scrollableElement: HTMLElement ) => {
+	const marginError = 3;
```
I think the `Error` suffix didn't help me understand what's this for 😅. Is `overflowThreshold` more accurate? Also, is this in pixels? A comment for posterity would be helpful.

cpapazoglou commented on plugins/woocommerce/client/blocks/tests/e2e/tests/product-gallery/inner-blocks/product-gallery-thumbnails/product-gallery-thumbnails.block_theme.spec.ts:
```diff
@@ -70,37 +70,4 @@ test.describe( 'Product Gallery Thumbnails block', () => {
 			).toBeVisible();
 		} );
 	} );
-
```
Let's add a test relevant to the new functionality.

cpapazoglou commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductGallery.php:
```diff
@@ -135,6 +135,10 @@ protected function render( $attributes, $content, $block ) {
 						'productId'         => $product_id,
 						'selectedImageId'   => $initial_image_id,
 						'userHasInteracted' => false,
+						'overflowTop'       => false,
+						'overflowBottom'    => false,
+						'overflowLeft'      => false,
+						'overflowRight'     => false,
```
Do you think that prefixing those with `thumbnails` would make sense?

kmanijak replied:
Yeah, context is "global" for whole Product Gallery while overflow is inner block specific. Good catch!

cpapazoglou commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductGalleryThumbnails.php:
```diff
@@ -75,50 +75,41 @@ protected function render( $attributes, $content, $block ) {
-		$remaining_thumbnails_count = $number_of_images - $number_of_thumbnails;
-		wp_interactivity_config( 'woocommerce/product-gallery', array( 'numberOfThumbnails' => $number_of_thumbnails ) );
-		// End of temporary solution.
+		$thumbnail_size   = str_replace( '%', '', $attributes['thumbnailSize'] ?? '20%' );
```
Similar to https://github.com/woocommerce/woocommerce/pull/56069/files#r1995705931
```suggestion
		$thumbnail_size   = str_replace( '%', '', $attributes['thumbnailSize'] ?? '33%' );
```

cpapazoglou commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/editor.scss:
```diff
@@ -0,0 +1,40 @@
+.wc-block-product-gallery-thumbnails {
```
Naive question, can we achieve the same result with a single `scss` file and not introduce this one?

kmanijak replied:
Probably we could but we should not. This is the pattern we use to separate Editor-specific styles from the frontend ones. Basically what's in `style.scss` will be included in Editor and frontend, while `editor.scss` will be applied in Editor only.

cpapazoglou commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/edit.tsx:
```diff
@@ -1,25 +1,78 @@
+		};
+	}, [ thumbnailSize ] ); // Re-run when thumbnailSize changes as it affects layout
+
+	const thumbnailSizeValue =
+		Math.min(
+			Math.max( Number( thumbnailSize.replace( '%', '' ) ), minSize ),
+			maxSize
+		) || defSize;
```
Isn't `minSize > thumbnailSize > maxSize` guaranteed from the block control settings?

kmanijak replied:
It is but only on blur. During the input any value is passed which causes a huge layout jumps. For example when you want to input `15`, when you remove everything, the value was a default (33), you input `1` and it goes to `1`, then `15`. Or if you accidentally put higher number then it's super big. I thought having some validation would be helpful.

I moved that logic to `onChange` in `plugins/woocommerce/client/blocks/assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/block-settings/index.tsx` though.

cpapazoglou replied:
Maybe switching back to `RangeControl` makes more sense and would avoid this complexity?

kmanijak replied:
I'd stick with `UnitControl` to keep it consistent across Editor. This is the control used to input sizes.

cpapazoglou commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-gallery/style.scss:
```diff
@@ -164,96 +174,83 @@ $dialog-padding: 20px;
 
-	@for $i from 3 through 8 {
-		&.wc-block-product-gallery-thumbnails--number-of-thumbnails-#{$i} {
+	@for $i from 10 through 50 {
```
I would add a comment that `10` and `50` are related to the `min`, `max` block values.

kmanijak committed:
Add changelog

kmanijak committed:
Update plugins/woocommerce/client/blocks/assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/block-settings/index.tsx

Co-authored-by: Harris Papazoglou <hrrsppzgl@gmail.com>

kmanijak committed:
Update plugins/woocommerce/client/blocks/assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/block-settings/index.tsx

Co-authored-by: Harris Papazoglou <hrrsppzgl@gmail.com>

kmanijak committed:
Update plugins/woocommerce/src/Blocks/BlockTypes/ProductGalleryThumbnails.php

Co-authored-by: Harris Papazoglou <hrrsppzgl@gmail.com>

kmanijak committed:
Rename context variables to include thumbanils context

kmanijak committed:
Add return types to util

kmanijak committed:
Move the thumbnails size validation logic to inspector controls

kmanijak committed:
Add styles comment

kmanijak committed:
Update context values

kmanijak committed:
Merge branch 'trunk' into add/56025

kmanijak committed:
Post merge change

kmanijak committed:
Bail early if scrollableElement is not HTMLELements

kmanijak committed:
Merge branch 'trunk' into add/56025

kmanijak committed:
Fix the instance check

kmanijak committed:
Show the horizontal scroll

kmanijak committed:
First E2E test

kmanijak committed:
Add test

kmanijak committed:
Add test for proper width

kmanijak committed:
Add test for scrolling

kmanijak committed:
Improve stability of test

kmanijak committed:
Add Safari specific styles

kmanijak committed:
Remove conflicting style

kmanijak committed:
Add vertical specific style

kmanijak committed:
Add vertical specific style in mobile layout

kmanijak committed:
Update E2E test to rely on actual sizes and not class

kmanijak committed:
Fix flaky test

kmanijak committed:
Fix flaky test

kmanijak commented:
@cpapazoglou , Thanks for thorough review! It's been a while but I hope I addressed everything from comments and our private conversations:
- bugs,
- code improvements
- added couple of tests (and it took me a while to fix flakiness)

I'd appreciate another round of review! 🙌

kmanijak commented:
✅ I tested the changes in Chrome, Firefox, Edge and Safari: desktop and mobile views.

kmanijak committed:
Merge branch 'trunk' into add/56025

cpapazoglou commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-gallery/inner-blocks/product-gallery-thumbnails/editor.scss:
```diff
@@ -0,0 +1,40 @@
+	// Editor uses :after pseudo-element to show outlines on hover or focus.
+	// This is a workaround to ensure the outline is visible even when the element is focused
+	// and we "override" :before pseudo-element so it's used as bottom and right gradient.
+	&::after {
+		opacity: 1 !important;
+		width: 100% !important;
+		height: 100% !important;
+		background: none !important;
```
Have you considered increasing specificity instead of `important`?

kmanijak replied:
Considering it's in Editor and it's strongly expected to bring those values to the "editor defaults" I didn't. What I mean is, in this particular case I don't see much value in seeking stronger specificity, there's no argue or condition in which those values should be overwritten by even stronger specificity.

cpapazoglou commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductGallery.php:
```diff
@@ -125,16 +125,20 @@ protected function render( $attributes, $content, $block ) {
+						'productId'                => $product_id,
+						'selectedImageId'          => $initial_image_id,
+						'userHasInteracted'        => false,
+						'thumbnailsOverflowTop'    => false,
+						'thumbnailsOverflowBottom' => false,
+						'thumbnailsOverflowLeft'   => false,
+						'thumbnailsOverflowRight'  => false,
```
Nitpick: Thinking of it again, I wonder if pushing those attributes to an object 
```
thumbnailsOverflow: {
 top:false,
 bottom: false.
 left: false,
 right: false
}
```

would be more readable and easier to pass in context.

kmanijak replied:
Actually it does make sense! https://github.com/woocommerce/woocommerce/pull/56069/commits/a5b1bb63ba98ce15644af8176595aaf16e185756

cpapazoglou reviewed (changes requested):
I have tested the various cases and found the following:

When switching to vertical and back to horizontal orientation `is-horizontal` class gets added which creates the following problems:

- in touch devices, thumbnails are not scrollable

- in Safari the layout is broken
![image](https://github.com/user-attachments/assets/8f101fcc-1b7a-4488-aa92-fbee2959123b)

This rule is breaking the above
![image](https://github.com/user-attachments/assets/8af068fc-b4f9-42c9-92cd-8ba6094ab6e9)

BTW, thanks for adding all those test cases 💪

kmanijak committed:
Fix the horizontal/vertical style application

kmanijak committed:
Fix styles

kmanijak committed:
Convert overflow state to an object

kmanijak commented:
Thanks @cpapazoglou ! I think now the layout is fixed. As discussed elsewhere, I removed the condition regarding `.is-horizontal` which means it's now possible to achieve such a layout but it's expected. Thumbnails obey the config from Gallery Area:

<img width="778" alt="image" src="https://github.com/user-attachments/assets/41e6ec9e-1a16-4ad6-8e25-dfbd09276003" />

I tested various combinations on Chrome, Firefox, Safari and Edge, both desktop and mobile. Right now seems like it's working 🙈

kmanijak committed:
Merge branch 'trunk' into add/56025

cpapazoglou reviewed (approved):
Thanks for rigorously iterating and addressing all feedback! Tested it in Chrome, Firefox, Safari, and in horizontal / vertical orientations - both in mobile and desktop.

It works as advertised! 🚢
